Friction options ab function

EstherHeck opened the pull request:
Pull Request Checklist. Please read and check each box with an X. Delete any part not applicable. Ask on the [forum](https://community.geodynamics.org/c/aspect) if you need help with any step.

*Describe what you did in this PR and why you did it.*

### Before your first pull request:

* [ ] I have read the guidelines in our [CONTRIBUTING.md](../blob/master/CONTRIBUTING.md) document.

### For all pull requests:

* [ ] I have followed the [instructions for indenting my code](../blob/master/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [ ] I have tested my new feature locally to ensure it is correct.
* [ ] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/master/tests/) directory.
* [ ] I have added a changelog entry in the [doc/modules/changes](../blob/master/doc/modules/changes) directory that will inform other users of my change.

naliboff committed:
add plastic damper

EstherHeck committed:
copied strain dependence and started modifying it for friction. declared friction entries

EstherHeck committed:
read friction parameters from input file

EstherHeck committed:
added function compute_dependent_friction_angle

EstherHeck committed:
strain dependence functions. Added composition mask for theta

EstherHeck committed:
deleted more strain dependence things

EstherHeck committed:
added function compute_edot_ii

EstherHeck committed:
prepared a compute_theta_reaction_terms function

EstherHeck committed:
added equation for dynamic friction

EstherHeck committed:
added equation for rate and state friction

EstherHeck committed:
comment about friction conversion for coefficient and radians

EstherHeck committed:
made the calculation of current_edot_ii part of compute_dependent_friction

EstherHeck committed:
put calculation of current_edot_ii back into its own function

EstherHeck committed:
added the calculation of cellsize

EstherHeck committed:
ran make indent

EstherHeck committed:
added comment on possible future friction option

EstherHeck committed:
incorporated some comment lines from Johns PR and conversion of dynamic friction angles into RAD

EstherHeck committed:
updated the dynamic friction parameters

EstherHeck committed:
cleaned up compute_dependent_friction a bit

EstherHeck committed:
deleted unnecessary const int i

EstherHeck committed:
deleted unnessary &out

EstherHeck committed:
deleted unnecessary comments in compute_dependent_friction

EstherHeck committed:
deleted unnesseary function

EstherHeck committed:
removed unnessary function

EstherHeck committed:
edited compute_theta_reaction_terms

EstherHeck committed:
made a few small changes and ran make indent. Now indenting is worse than before

EstherHeck committed:
found missing } , ran make indent, but indenting didnt change

EstherHeck committed:
found the line that caused the trouble and ran make indent

EstherHeck committed:
added a friction_options.h file

EstherHeck committed:
updated include and def statements

EstherHeck committed:
added input parameters in .h

EstherHeck committed:
tackled the first compile errors

EstherHeck committed:
tackled some more compile errors

EstherHeck committed:
tackled some more compile errors

EstherHeck committed:
ran make indent

EstherHeck committed:
added a get friction_dependence_mechanism function

EstherHeck committed:
added enum in .h which solved some errors

EstherHeck committed:
removed more compile errors

EstherHeck committed:
only one compile error left

EstherHeck committed:
calling compute_dependent_friction_angle in evaluate and additional outputs in visco_plastic.cc

EstherHeck committed:
added comments where I need to call functions for theta

EstherHeck committed:
modified theta_composition_mask and call it from visco_plastic

EstherHeck committed:
calling theta reaction terms from visco_plastic

EstherHeck committed:
include friction_options.h in visco_plastic.h

EstherHeck committed:
changed option none to independent. declared friction_options in visco_plastic

EstherHeck committed:
some comment

EstherHeck committed:
added an n evaluation points i to some function, but not cleaned up

EstherHeck committed:
outsourced computation of current_edot_ii from compute_dependent_friction

EstherHeck committed:
cleaned up compute_dependent_friction_angle

EstherHeck committed:
solved the second last error on how to hand over composition

EstherHeck committed:
added a comment about remaining error

EstherHeck committed:
handing over current_cell in is_yielding and _compute_isostrain_viscosities

EstherHeck committed:
added friction parameters in enter subsection of visco_plastic but got an error

EstherHeck committed:
added friction parameters in enter subsection of visco_plastic but got an error

EstherHeck committed:
added parsing of friction_option parameters in two places and resolved compile and runtime errors

EstherHeck committed:
changed critical slip distance to only one double instead of material dependent

EstherHeck committed:
runtime error: segmentation fault occuring in the beginning of time step 1 in line 328 where strain_rate is handed over

EstherHeck committed:
ran make indent

EstherHeck committed:
untangling the compute_edot_ii function

EstherHeck committed:
added a function that computes current_edot_ii

EstherHeck committed:
corrected the strain_rate passing

EstherHeck committed:
removed claculate_viscoelastic_strain_rate from elasticity and put it into utilities

EstherHeck committed:
added elastic shear moduli as an input for compute_current_edot_ii

EstherHeck committed:
removed const behind calculate_viscoelastic_strain_rate to get rid of an error

EstherHeck committed:
added missing template <int dim>

EstherHeck committed:
added elastic timestep as an inout for comput_current_edot_ii and viscoelastic_strain_rate-invariant

EstherHeck committed:
added missing elastic_rheology.

bobmyhill committed:
added composite_viscous rheology plugin

EstherHeck committed:
wrote explicit instantiation

EstherHeck committed:
rebased with master

bobmyhill committed:
update

EstherHeck committed:
adapting function inputs, but still errors left

EstherHeck committed:
still not compiling, am confused inside compute_theta_reaction_terms

bobmyhill committed:
allowed phase transitions in peierls

bobmyhill committed:
shortened existing prm

bobmyhill committed:
added new test

bobmyhill committed:
added changelog

bobmyhill committed:
addressed initial review comments

EstherHeck committed:
moved compute_edot_ii from Utilities into MaterialUtilities

EstherHeck committed:
ran make indent

EstherHeck committed:
adjusted the call of compute_current_edot_ii to MaterialUtilities

EstherHeck committed:
ran make indent

EstherHeck committed:
removed unnecessary input for function

EstherHeck committed:
added parameter theta_in_use

EstherHeck committed:
added function compute_theta

EstherHeck committed:
included Daniels changes for elastic timestep handling

EstherHeck committed:
added a function compute_plastic_strain_rate in drucker_prager from Johns PR 3473 and missing bits for elastic time step

EstherHeck committed:
calling the new compute_plastic_strain_rate function, but got errors

EstherHeck committed:
outcommented the line where I call the plastic_strain_rate function, as too much of Johns PR is needed here, but not tested yet

tjhei committed:
add "repeat on cutback" scheme, add test

EstherHeck committed:
removed 0.03

bobmyhill committed:
addressed changes proposed at user meeting

bobmyhill committed:
added new composite viscoplastic rheology files

EstherHeck committed:
elastic stress averaging is now possible without fixed elastic time step

EstherHeck committed:
added brackets and dots according to Annes advice

EstherHeck committed:
ran make indent

EstherHeck committed:
added the effective friction fator as an input

EstherHeck committed:
added const where possible

EstherHeck committed:
trying to make a and b a function of depth

EstherHeck committed:
ran make indent

EstherHeck committed:
got rid of some compile errors

EstherHeck committed:
added AandBSource in header file

EstherHeck committed:
ran make indent

EstherHeck committed:
put in print statements to check if a and b are read in correctly

EstherHeck committed:
made one subsection for a and b function each

EstherHeck committed:
tried to have a function for each composition as in initialComposition Function, but got error

EstherHeck committed:
two errors left about line 436 in friction_options.cc

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,198 @@
+    {
+      /**
+       * Enumeration for selecting which type of friction dependence to use.
+       * For independent, internal angle of friction is used.
```
Please clarify here, e.g. for the type 'independent', the user-supplied internal angle of friction is used.

What about strain weakening, this is handled elsewhere? But this would also affect the friction angle right?

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,198 @@
+          /**
+           * A function that returns if the state variable theta is used.
+           */
+          bool get_theta_in_use() const;
```
We usually name such a bool `use_theta`, the getter function would then be called `get_use_theta`.

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,198 @@
+        private:
+
+          FrictionDependenceMechanism friction_dependence_mechanism;
+
```
rm empty line

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,198 @@
+           * Objects for computing plastic stresses, viscosities, and additional outputs
+           */
+          Rheology::DruckerPrager<dim> drucker_prager_plasticity;
+          /**
```
add empty line before comment. In the header file, the format is one line between units.

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,198 @@
+          Rheology::StrainDependent<dim> strain_rheology;
+
+          /**
+          * dynamic friction input parameters
```
Please use capitals to start the comments to keep the formatting uniform, and give one comment for each parameter.

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,198 @@
+          double dynamic_characteristic_strain_rate;
+          double dynamic_friction_smoothness_exponent;
+
+          /**
```
Align and provide comment for each parameter.

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,198 @@
+          */
+          std::vector<double> effective_friction_factor;
+          double critical_slip_distance;
+          std::vector<double> steady_state_strain_rate;
```
Note to self to come back to this: why is this a vector?

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,198 @@
+          };
+
+          /**
+           * Currently chosen source for the viscosity.
```
-> for the depth dependency of the friction coefficient.

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,198 @@
+
+          /**
+           * An enum to describe where the depth dependency of the a and b value for
+           * rate and state friction is coming from.
```
Can you describe the options? 

What other options do you plan to add in the future? (Or, do you need the enum?)

EstherHeck replied:
You are right, I do not need the enum. It should always be function. How do correctly write this?

anne-glerum replied:
If it's always a function, then there is no need for this parameter, and you just implement the function.

EstherHeck replied:
ok cool, done. How do I set the default for the function?

anne-glerum commented on include/aspect/material_model/utilities.h:
```diff
@@ -459,6 +459,31 @@ namespace aspect
+
+      /**
+       * Computes the current_edot_ii and current_stress, with the option for
+       * viscous and viscoelastic contribution.
```
-> contributions

anne-glerum commented on source/material_model/rheology/drucker_prager.cc:
```diff
@@ -50,6 +50,39 @@ namespace aspect
+        const double stress_inv_part = 1. / (std::sqrt(3.0) * (3.0 + sin_phi));
+
+        // Initial yield stress (no stabilization terms)
+        double yield_stress = ( (dim==3)
```
-> const double

anne-glerum commented on source/material_model/rheology/drucker_prager.cc:
```diff
@@ -50,6 +50,39 @@ namespace aspect
+        // Calculate the plastic strain rate, which is needed to calculate the additional stress
+        // associated with the plastic damper.
+        const double total_stress = ( yield_stress + ( 2. * damper_viscosity * effective_strain_rate ) ) /
+                                    ( 1 + ( damper_viscosity / pre_yield_viscosity ) );
```
-> 1.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
```
rm empty line

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+                           "\\item ``state dependent friction'': A state variable theta is introduced. Method "
+                           "is taken from Sobolev and Muldashev 2017. ....."
+                           "\n\n"
+                           "\\item ``default'': No rate or state dependence of the friction angle is applied. ");
```
`default` does the same as `none` right? Then I'd make `none` the default and remove `default`.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+      {
+        prm.declare_entry ("Friction dependence mechanism", "default",
+                           Patterns::Selection("none|dynamic friction|state dependent friction|default"),
+                           "Whether to apply a rate and/or state dependence of the friction angle. This can "
```
From this entry it is not clear that:
1) dynamic friction = rate dependent
2) how to apply rate _and_ state dependence, which is written as an option here (`rate and/or state`)

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+
+        // Plasticity parameters
+        /*should I do this or just read in the internal anlges of friction directly? */
+        /*drucker_prager_parameters = drucker_prager_plasticity.parse_parameters(this->n_compositional_fields()+1,
```
Good question. Although you parse more parameters than necessary this way (in this case only 2), using 'parse_parameters' does ensure that the reading in of the same parameters in different places is done in the same way, for example the conversion from degrees to radians for the internal friction angle. I think this is preferable.

EstherHeck replied:
In the process of writing this module I decided to pass the internal angle of friction from visco_plastic.cc to friction_options as it is already modified by the weakening factors at that point. So I guess I just remove those comment lines?

EstherHeck replied:
or I move the weakening factors to after compute_dependent_friction_angle in line 338 ?

anne-glerum replied:
If you can avoid doing additional work by getting the strain-weakened internal angle of friction from visco_plastic, than I'd do that.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+                           Patterns::Double (0),
+                           "The characteristic strain rate value, where the angle of friction takes the middle "
+                           "between the dynamic and the static angle of friction. When the effective strain rate "
+                           "in a cell is very high the dynamic angle of friction is taken, when it is very low "
```
-> is very high, the dynamic

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+                           Patterns::List(Patterns::Double(0)),
+                           "List of dynamic angles of internal friction, $\\phi$, for background material and compositional "
+                           "fields, for a total of N+1 values, where N is the number of compositional fields. "
+                           "For a value of zero, in 2D the von Mises criterion is retrieved. "
```
rm this line

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+                           "List of dynamic angles of internal friction, $\\phi$, for background material and compositional "
+                           "fields, for a total of N+1 values, where N is the number of compositional fields. "
+                           "For a value of zero, in 2D the von Mises criterion is retrieved. "
+                           "Dynamic angles of friction are used for calculation when the effective strain rate in a cell "
```
for the calculation of what? Please rephrase.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+        prm.declare_entry ("Dynamic friction smoothness exponent", "1",
+                           Patterns::Double (0),
+                           "An exponential factor in the equation for the calculation of the friction angle "
+                           "when a static and a dynamic friction angle is specified. A factor =1 is equivalent "
```
-> are specified
-> A factor of 1 returns the equation to equation 13

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+                           "Units: none.");
+
+        // rate and state parameters
+        prm.declare_entry ("Rate and state parameters a", "None",
```
The parameter name is not clear. From the current name I would expect to prescribe a scalar value. Also as it relates to one parameter, parameter a, I'd not write 'parameters'. Please rename.

EstherHeck replied:
I see what you mean. Actually this can only be "Function" once I delete the enum, which I do not need.  So how do I set the default function to just be 1. or 0. everywhere or something similar?

anne-glerum replied:
I think this parameter can go then. In the `Functions::ParsedFunction<dim>::declare_parameters(prm,1)`, the function expression will be set to zero by default.

anne-glerum replied:
BTW, you could also use one function with two components, one for a and one for b. Not sure if it simplifies things much.

EstherHeck replied:
cool. right now I am using this parameter also in:
       if ( prm.get("Rate and state parameter a") == "Function" )
          {
            prm.enter_subsection("Rate and state parameter a function");

what should I do here instead then? Just leave the if out?

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+                           Patterns::Selection("Function|None"),
+                           "Method that is used to specify how the a and b should vary with depth."
+                           "The rate and state parameter a is the rate dependency. "
+                           "The rate and state parameter b is the state dependency. Positive (a-b) is velocity"
```
add space before last "
-> (a-b) corresponds to velocity

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+                           "Method that is used to specify how the a and b should vary with depth."
+                           "The rate and state parameter a is the rate dependency. "
+                           "The rate and state parameter b is the state dependency. Positive (a-b) is velocity"
+                           " strengthening, negative (a-b) is velocity weakening. "
```
remove first space

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+                           "The rate and state parameter a is the rate dependency. "
+                           "The rate and state parameter b is the state dependency. Positive (a-b) is velocity"
+                           " strengthening, negative (a-b) is velocity weakening. "
+                           "Units: none");
```
dot after none

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+                           "The rate and state parameter b is the state dependency. Positive (a-b) is velocity"
+                           " strengthening, negative (a-b) is velocity weakening. "
+                           "Units: none");
+        prm.declare_entry ("Rate and state parameters b", "None",
```
Same comments as for a

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+        // rate and state parameters
+        prm.declare_entry ("Rate and state parameters a", "None",
+                           Patterns::Selection("Function|None"),
+                           "Method that is used to specify how the a and b should vary with depth."
```
This entry only specifies how a should vary with depth.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+                           " strengthening, negative (a-b) is velocity weakening. "
+                           "Units: none");
+
+        prm.enter_subsection("a function");
```
'Rate and state parameter a function'? 'a function' sounds a little weird when you don't realize the parameter is called 'a' :)

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+                           "A number that is multiplied with the coefficient of friction to take into "
+                           "account the influence of pore fluid pressure. This makes the friction "
+                           "coefficient an effective friction coefficient as in Sobolev and "
+                           "Muldashev 2017. "
```
-> (2017)
Even better would be to add the full reference to aspect reference list (doc/manual/manual.bib) and insert the citation key here with \\cite{}.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+        prm.declare_entry ("Critical slip distance", "0.01",
+                           Patterns::List(Patterns::Double(0)),
+                           "The critical slip distance in rate and state friction. Used to calculate the state "
+                           "variable theta.   "
```
rm additional spaces at the end and add . after `m` in the next line. Can you add how it is used to come up with a velocity?

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+        prm.declare_entry ("Steady state strain rate", "1e-14",
+                           Patterns::List(Patterns::Double(0)),
+                           "Arbitrary strain rate at which friction equals the reference friction angle in "
+                           "rate and state friction. "
```
Can you add some text about how to choose this and how it should/does/does not relate to the `dynamic characteristic strain rate`? Why is it called `Steady state`?

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+                           Patterns::List(Patterns::Double(0)),
+                           "Arbitrary strain rate at which friction equals the reference friction angle in "
+                           "rate and state friction. "
+                           "Units: $1/s$");
```
-> $."

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+                           "rate and state friction. "
+                           "Units: $1/s$");
+      }
+
```
This for everywhere in this file: 3 empty lines between class functions.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+          friction_dependence_mechanism = dynamic_friction;
+        else if (prm.get ("Friction dependence mechanism") == "state dependent friction")
+          friction_dependence_mechanism = state_dependent_friction;
+        else if (prm.get ("Friction dependence mechanism") == "default")
```
remove `default` option, see header file

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+        // Rate and state friction parameters
+        if (friction_dependence_mechanism == state_dependent_friction)
+          {
+            AssertThrow(this->introspection().compositional_index_for_name("theta"),
```
`compositional_index_for_name("theta")` will return an integer, unless theta doesn't exist, then it will throw an error. In c++, anything that is not zero is considered true; if theta has index zero, this assert will throw an exception...
You could use `compositional_name_exists` instead.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+                        ExcMessage("Material model with rate-and-state friction only works "
+                                   "if there is a compositional field that is called theta."));
+          }
+
```
rm empty line

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+
+        if ( prm.get("Rate and state parameters a") == "Function" )
+          {
+            a_and_b_source = Function;
```
this parameter name is confusing and possibly leads to wrong behavior down the line. I'd have a separate parameter for a and for b. So `a_source = Function`.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+            {
+              try
+                {
+                  //rate_and_state_parameter_a_function.parse_parameters(prm);
```
rm

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+
+        if ( prm.get("Rate and state parameters b") == "Function" )
+          {
+            a_and_b_source = Function;
```
See my comment above. Here `a_and_b_source` is overwritten.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+            {
+              try
+                {
+                  //rate_and_state_parameter_b_function.parse_parameters(prm);
```
rm

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+          }
+      }
+
+
```
Please stick to the general order of functions, where we declare and parse parameters at the end of the file.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+          {
+            case independent:
+            {
+              /* I guess here I need to say that current friction is simply the internal angle of friction. If not it will return 0.0 as its value?  */
```
at the moment it does not change the value that was in `current_friction`

EstherHeck replied:
so it is fine like this?

anne-glerum replied:
If you don't want the friction angle to change in this function when it is `independent`, then yes (which I think is the case here). So you can remove the comment and keep `break`.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+              // const double mu  = mu_d[i] + (mu_s[i] - mu_d[i]) / ( (1 + strain_rate_dev_inv2/reference_strain_rate) );
+              // which is the following using the variables in this material_model. Although here the dynamic friction coefficient
+              // is directly specified. The coefficient of friction is the tangent of the internal angle of friction.
+              // Furthermore a smoothness coefficient is added, which influences if the friction vs strain rate curve is rather
```
influences if -> determines whether

anne-glerum replied:
Please also write out the equation with the smoothness exponent.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+            }
+            case dynamic_friction:
+            {
+              // The dynamic characteristic strain rate is used to see if dynamic or static angle of internal friction should be used.
```
it's not either dynamic or static right? But a gradual transition from one to the other.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+            }
+            case state_dependent_friction:
+            {
+              //cellsize is needed for theta and the friction angle
```
Keep the same format of the comments (e.g. `// cellsize`) throughout the file.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+                  const double rate_and_state_parameter_a = calculate_depth_dependent_a_and_b(position,j).first;
+                  const double rate_and_state_parameter_b = calculate_depth_dependent_a_and_b(position,j).second;
+
+                  // calculate effective friction according to equation (4) in Sobolev and Muldashev 2017;
```
-> (2017)

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+
+                  // calculate effective friction according to equation (4) in Sobolev and Muldashev 2017;
+                  // effective friction is calculated by multiplying the friction coefficient with 0.03 = (1-p_f/sigma_n)
+                  // their equation is for friction coefficient, while ASPECT takes friction angle in RAD, so conversion with tan/atan()
```
RAD -> radians
Please also add punctuation when writing full sentences.
Is the equations written down in a comment somewhere? If not, this would be a good place.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+                  // their equation is for friction coefficient, while ASPECT takes friction angle in RAD, so conversion with tan/atan()
+                  current_friction = atan(effective_friction_factor[j] * tan(current_friction)
+                                          + rate_and_state_parameter_a
+                                          * log((current_edot_ii * cellsize ) / steady_state_strain_rate[j]) + rate_and_state_parameter_b
```
Can you put the log term on the previous line and rate_and_state_parameter_b on the next? I think that is more clear to understand.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+                    const double current_edot_ii,
+                    const double cellsize) const
+      {
+        // equation (7) from Sobolev and Muldashev 2017. Though here I had to add  "- theta_old"
```
This comment seems out of place as you're computing the current theta, not the change in theta.
Can you add the formula?

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+          {
+            for (unsigned int j=0; j < volume_fractions.size(); ++j)
+              {
+                // q war n_evaluation_points
```
Update to English or remove

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+        //cellsize is needed for theta and the friction angle
+        double cellsize = 1;
+        if (in.current_cell.state() == IteratorState::valid)
+          {
```
Can you add a comment why you use the extent in the X-direction?

EstherHeck replied:
at the moment this is fine for what I do, because the cells are cubic and motion is in y-direction. But I remember that we discussed that it should probably rather be cell size in the direction of most movement. Unfortunately I do not remember how I can specify this. Do you have an idea?

anne-glerum replied:
To get the length of the element in the slip direction, you'd have to
1. determine whether you're in the fault
2. if so, get the velocity vector
3. get vectors along two non-parallel edges of the element (I'm foreseeing mesh deformation here)
4. from this compute the length of the element along the velocity vector.
For now this is too much. If you can add a comment here that explains your choice, then it's fine, e.g.
`For now, the used cells are non-deforming squares, so the edge length in the x-direction is representative of the cell size. TODO as the cell size is used to compute the slip velocity as cell_size * strain_rate, come up with a better representation of the slip length.`

EstherHeck replied:
thanks!

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+            cellsize = in.current_cell->extent_in_direction(0);
+          }
+
+        if (this->simulator_is_past_initialization() && this->get_timestep_number() > 0 && in.requests_property(MaterialProperties::reaction_terms) && in.current_cell.state() == IteratorState::valid)
```
Please break this into two lines

EstherHeck committed:
somethign changed after holidays

EstherHeck committed:
still unsuccessfully trying to access coordinate_system

anne-glerum commented on include/aspect/material_model/rheology/friction_options.h:
```diff
@@ -0,0 +1,198 @@
+           * Function to calculate depth-dependent a and b values for state dependent friction.
+           */
+          std::pair<double,double>
+          calculate_depth_dependent_a_and_b(const Point<dim> &position, const int j) const;
```
The usual order is class functions and then parameters

EstherHeck replied:
ok, I thought it would be convenient to have everything concerning a and b at one place in the script. What do you think?

anne-glerum replied:
For me, this order of functions and then parameters takes precedence. It helps when you're reading code when your expectations are met regarding code structure. That said, looking at some ASPECT header files, this order is not always kept. So it's up to you.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+      std::pair<double,double>
+      FrictionOptions<dim>::calculate_depth_dependent_a_and_b(const Point<dim> &position, const int j) const
+      {
+        if ( a_and_b_source == Function )
```
As said before, the input parameters ask for a source for both a and b, while here you assume there is one source for both.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+        if ( a_and_b_source == Function )
+          {
+            Utilities::NaturalCoordinate<dim> point =
+              this->get_geometry_model().cartesian_to_other_coordinates(position, coordinate_system);
```
What coordinate system is set here? From the comments and function name, it seems like you only want depth as a function variable.

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+      FrictionOptions<dim>::
+      get_theta_in_use() const
+      {
+        bool theta_in_use =false;
```
-> = false

EstherHeck replied:
is it then also ->=true when I update it? Like this: 

bool use_theta ->=false;
if (get_friction_dependence_mechanism() == rate_and_state_dependent_friction)
    use_theta ->=true;

anne-glerum replied:
I should have explained better what I meant. With `->` I mean please change into whatever comes next. In this case this means a space between `=` and `false`.

EstherHeck replied:
o true, I understood that in the other comments ;)

anne-glerum commented on source/material_model/rheology/friction_options.cc:
```diff
@@ -0,0 +1,495 @@
+              current_friction = std::atan (mu);
+              break;
+            }
+            case state_dependent_friction:
```
a and b are called `rate and state dependent parameters`, but they are used in the `state_dependent_friction` option only. Can you rethink the names?

anne-glerum reviewed (changes requested):
Hi @EstherHeck, great progress, nicely implemented! Most of the comments are stylistic. Also, I might have commented on things that somebody else implemented. Implementation-wise, you have to decide whether you want a separate a_and_b_source for a and b, or just one for both. And whether you want the function for a and b to only be depth dependent, or to allow for other coordinate systems as well. I wouldn't mind having another look after the comments are addressed now that I've gotten an overview.

EstherHeck committed:
adressed most of Annes comments for friction_options.h

EstherHeck committed:
addressed more of Annes comments, mostly about naming and explanation/documentation

EstherHeck committed:
added empty lines between class functions

EstherHeck committed:
addressed all comments from Anne that did not involve swapping places of code snippets

EstherHeck committed:
inserted latex citations

EstherHeck committed:
moved declare and parse parameters to end of file

EstherHeck committed:
added RSF equation and aging law to description

EstherHeck committed:
added . for double

EstherHeck committed:
most of Annes comments adressed now. a and b function updated. still compile error left about coordinates

EstherHeck committed:
deleted a comment, changed order of function and params in .h

EstherHeck committed:
removed if == function statement, added comment, ran make indent, included .h file so now it compiles

EstherHeck committed:
changed \frac etc to \\frac in latex documentation

EstherHeck committed:
something

EstherHeck committed:
deleted couts, abFunction now properly working

gassmoeller committed:
Allow non-uniform depth average layers

EstherHeck committed:
convert minimum time step to years if needed

EstherHeck committed:
deleted ;

MFraters committed:
refactor set_default_assemblers to fix the Newton solver for heatflow boundaries.

MFraters committed:
add changelog entry for Newton solver fix for heat flux boundaries

gassmoeller committed:
Update dockerfile

gassmoeller committed:
Merge pull request #3865 from gassmoeller/diffusion_follow_up

Minor improvements to surface diffusion

gassmoeller committed:
Update trusted CI users

gassmoeller committed:
Merge pull request #3870 from MFraters/fix_Newton_solver_with_heat_flux_boundaries

refactor set_default_assemblers to fix the Newton solver for heatflow boundaries

gassmoeller committed:
Merge pull request #3860 from tjhei/timestepping_min

TimeStepping: implement a minimum time step size

tjhei committed:
Merge pull request #3872 from gassmoeller/update_trusted_developers

Update trusted CI users

bobmyhill committed:
rebase, updated test

EstherHeck committed:
ran make indent

jdannberg committed:
Merge pull request #3825 from naliboff/gerya_2019_vep_benchmark

add new vep benchmark

kchotalia committed:
default make_periodicity_constraints in interface

EstherHeck committed:
timesteps can now not be repeated with values smaller than the minimum timestep

gassmoeller committed:
Hand over DoFHandler as parameter

gassmoeller committed:
Minor fixes

gassmoeller committed:
Unify postprocessor sceen output

gassmoeller committed:
Merge pull request #3841 from bobmyhill/add_phases_peierls

Add phases to Peierls

gassmoeller committed:
Merge pull request #3874 from gassmoeller/update_gravity_postprocessor

Unify postprocessor sceen output

naliboff committed:
add benchmark

jdannberg committed:
Merge pull request #3779 from kchotalia/quarter_shell_spherical_periodic

Quarter shell spherical periodic

EstherHeck committed:
moved min out of the loop

EstherHeck committed:
added a max() to consider min dt when repeating a time step

EstherHeck committed:
put the max() into an if

EstherHeck committed:
removed :

gassmoeller committed:
Address comments

gassmoeller committed:
Add newline

tjhei committed:
Merge pull request #3871 from gassmoeller/update_docker_image

Update dockerfile

EstherHeck committed:
added asserts and todos for asserts to prevent negative friction angles

EstherHeck committed:
added the parameter use_radiation_damping

EstherHeck committed:
ran make indent, added citation

EstherHeck committed:
added radiation damping in the cohesions

gassmoeller committed:
Add postprocessor for principal stresses

gassmoeller committed:
Add changelog

EstherHeck committed:
a half compiling code where radiation damping is taken into account in both drucker-prager files

EstherHeck committed:
radiation damping available for visco_plastic material model only. Compiling fine

EstherHeck committed:
yield stress plus radiation damping

EstherHeck committed:
yield stress minus radiation damping

tjhei committed:
update parameters

jdannberg committed:
Merge pull request #3866 from gassmoeller/add_scaled_viscosity_profile

Allow non-uniform depth thicknesses for lateral averaging

gassmoeller committed:
Merge pull request #3879 from tjhei/maintenance_nov2020

update parameters

naliboff committed:
add flag for plastic damper

naliboff committed:
add test

EstherHeck committed:
assertthrow if theta is <= 0

EstherHeck committed:
assertthrow if a or b is negative

bangerth committed:
Make error messages more visible when compiling manual.tex.

bobmyhill committed:
added test

bobmyhill committed:
added changelog, indented

EstherHeck committed:
radiation damping now part of current stress instead of yield stress

EstherHeck committed:
current-edot-ii modified by current-stress if radiation damping used

EstherHeck committed:
radiation damping also considered in plastic outputs

tjhei committed:
Merge pull request #3880 from bangerth/make

Make error messages more visible when compiling manual.tex.

jdannberg committed:
add depth averaging for compositional field mass

jdannberg committed:
update tests

kchotalia committed:
periodic tracers in quarter shell, adding rotation

bobmyhill committed:
extended docstrings

ljhwang committed:
adding Assanelli et al to publications

jdannberg committed:
address comments

gassmoeller committed:
Merge pull request #3881 from jdannberg/composition_mass_averaging

add depth averaging for compositional field mass

tjhei committed:
Merge pull request #3882 from kchotalia/particle_periodic_quarter_shell

Periodic particles in quarter shell

tjhei committed:
Merge pull request #3883 from ljhwang/update_citing_aspect

adding Assanelli et al to publications

naliboff committed:
address comments

bobmyhill committed:
refactor elastic viscosity

bobmyhill committed:
removed factor two in elasticity strain rate calculation

tjhei committed:
fix undefined behavior in visco_plastic

tjhei committed:
add missing override

tjhei committed:
add missing instantiation

internal::GPlatesLookup is used in boundary_velocity_residual_statistics
and clang correctly complains that we don't explicitly instantiate the
template class. Fix this.

MFraters committed:
Merge pull request #3890 from tjhei/missing_override

add missing override

tjhei committed:
update clang-tidy settings and run script

- update to new clang-tidy-10 settings
- disable unity/pch support in script

tjhei committed:
apply fixes

undo boost bind change

gassmoeller committed:
Merge pull request #3891 from tjhei/missing_gplates_instantiation

add missing instantiation

gassmoeller committed:
Merge pull request #3889 from tjhei/undefined_beh_visco

fix undefined behavior in visco_plastic

tjhei committed:
remove virtual when using override

follow-up to #3890

bangerth committed:
Merge pull request #3892 from tjhei/override_no_virtual

remove virtual when using override

bobmyhill committed:
updated test output

tjhei committed:
update test results

update test

anne-glerum committed:
Fix sign of shear stress

bobmyhill committed:
addressed comments

naliboff committed:
Merge pull request #3886 from bobmyhill/refactor_elastic_viscosity

Refactor viscoelastic viscosity

bobmyhill committed:
modified comment

gassmoeller committed:
Update test output

bobmyhill committed:
extended comment on elastic strain rate

naliboff committed:
Merge pull request #3888 from bobmyhill/factor_two_elasticity

removed factor two in elasticity strain rate calculation

anne-glerum committed:
Fix sign surface stress plus comments

anne-glerum committed:
Update descriptions

anne-glerum committed:
Add changelog

anne-glerum committed:
Indent

EstherHeck committed:
added cout for theta or friction angle <= 0

EstherHeck committed:
assert Throw if a solver scheme with more than one advection  is chosen together with rate and state friction

naliboff committed:
Merge pull request #3878 from gassmoeller/principal_stress_components

Principal stress visulization output

naliboff committed:
add changelog

naliboff committed:
update maintainer and developer list

tjhei committed:
allow bind

tjhei committed:
address comments

anne-glerum committed:
Update test comment

anne-glerum committed:
Update tests

naliboff committed:
Merge pull request #3897 from anne-glerum/fix_stress

Fix stress

MFraters committed:
Merge pull request #3895 from tjhei/clang-tidy-update

clang-tidy update

MFraters committed:
rename Steinberger material model variable.

MFraters committed:
Prevent a warning in release mode of an unused variable in principle stress visualization

naliboff committed:
Merge pull request #3901 from MFraters/prevent_warning_in_principle_stress_visualization

Prevent a warning in release mode of an unused variable in principle stress visualization

naliboff committed:
Merge pull request #3900 from MFraters/rename_Steinberger_material_model_variable

rename Steinberger material model variable.

tjhei committed:
Merge pull request #3899 from naliboff/update_developers_maintainers

Update README maintainer and developer lists

naliboff committed:
update variable name

naliboff committed:
change additional variable names

bobmyhill committed:
Merge pull request #3902 from naliboff/rename_variable

Rename strain rate variable

EstherHeck committed:
deleted two unnecessary lines in evaluate-compute-theta-reaction-terms and added lots of temporal print statements

EstherHeck committed:
put j-loop around reaction terms as it is in calculate-isostrain-viscosities

gassmoeller committed:
Refactor some parameters

naliboff committed:
Merge pull request #7 from gassmoeller/plastic_damper

Refactor some parameters

EstherHeck committed:
fixed the triggering of theta assert

EstherHeck committed:
cleaned up some usage of elastic and strain rheology in frictionOptions

EstherHeck committed:
radiation damping only possible with elasticity

EstherHeck committed:
use average_shear_moduli in compute_theta_reaction_terms

EstherHeck committed:
theta_reaction terms insode if(use_elasticity) statement

EstherHeck committed:
Revert "cleaned up some usage of elastic and strain rheology in frictionOptions"

This reverts commit a6f89db2794bf9d8007a6071e0b103f511bd4cb7.

EstherHeck committed:
cleaned up some includes in friction_options.h

EstherHeck committed:
comment about how to get real density for radiation damping

naliboff committed:
Merge pull request #3823 from bobmyhill/composite_viscous

New: Composite viscous creep rheology

bobmyhill committed:
Merge pull request #3473 from naliboff/plastic_damper

Thanks @naliboff and @gassmoeller! Now merged.

EstherHeck committed:
merged with master and cleaned up afterwards

EstherHeck committed:
access plastic_strain_rate computation in visco_plastic.cc

EstherHeck committed:
use plastic strain rate for friction angle and radiation damping

EstherHeck committed:
comment about which strain rate is used for friction

EstherHeck committed:
set lower bound to plastic strain rate for RSF to min strainrate
